File: jbake-core/src/main/java/org/jbake/model/DocumentModel.java
package org.jbake.model;

import org.jbake.app.DBUtil;

import java.util.Date;

public class DocumentModel extends BaseModel {

    public static DocumentModel createDefaultDocumentModel() {
        DocumentModel documentModel = new DocumentModel();
        documentModel.setCached(true);
        documentModel.setRendered(false);
        return documentModel;
    }

    public String getBody() {
        return (String) get(ModelAttributes.BODY);
    }

    public void setBody(String body) {
        put(ModelAttributes.BODY, body);
    }

    public Date getDate() {
        return (Date) get(ModelAttributes.DATE);
    }

    public void setDate(Date date) {
        put(ModelAttributes.DATE, date);
    }

    public String getStatus() {
        if (containsKey(ModelAttributes.STATUS)) {
            return (String) get(ModelAttributes.STATUS);
        }
        return "";

    }

    public void setStatus(String status) {
        put(ModelAttributes.STATUS, status);
    }

    public String getType() {
        if (containsKey(ModelAttributes.TYPE)) {
            return (String) get(ModelAttributes.TYPE);
        }
        return "";
    }

    public void setType(String type) {
        put(ModelAttributes.TYPE, type);
    }

    public String[] getTags() {
        return DBUtil.toStringArray(get(ModelAttributes.TAGS));
    }

    public void setTags(String[] tags) {
        put(ModelAttributes.TAGS, tags);
    }

    public String getSha1() {
        return (String) get(ModelAttributes.SHA1);
    }

    public void setSha1(String sha1) {
        put(ModelAttributes.SHA1, sha1);
    }

    public String getSourceuri() {
        return (String) get(ModelAttributes.SOURCE_URI);
    }

    public void setSourceUri(String uri) {
        put(ModelAttributes.SOURCE_URI, uri);
    }

    public String getRootPath() {
        return (String) get(ModelAttributes.ROOTPATH);
    }

    public void setRootPath(String pathToRoot) {
        put(ModelAttributes.ROOTPATH, pathToRoot);
    }

    public boolean getRendered() {
        return (boolean) getOrDefault(ModelAttributes.RENDERED, false);
    }

    public void setRendered(boolean rendered) {
        put(ModelAttributes.RENDERED, rendered);
    }

    public String getFile() {
        return (String) get(ModelAttributes.FILE);
    }

    public void setFile(String path) {
        put(ModelAttributes.FILE, path);
    }

    public String getNoExtensionUri() {
        return (String) get(ModelAttributes.NO_EXTENSION_URI);
    }

    public void setNoExtensionUri(String noExtensionUri) {
        put(ModelAttributes.NO_EXTENSION_URI, noExtensionUri);
    }

    public String getTitle() {
        return (String) get(ModelAttributes.TITLE);
    }

    public void setTitle(String title) {
        put(ModelAttributes.TITLE, title);
    }

    public Boolean getCached() {

        Object value = get(ModelAttributes.CACHED);
        if ( value instanceof String ) {
            return Boolean.valueOf((String) value);
        } else {
            return (Boolean) value;
        }
    }

    public void setCached(boolean cached) {
        put(ModelAttributes.CACHED, cached);
    }

    public void setNextContent(DocumentModel nextDocumentModel) {
        put(ModelAttributes.NEXT_CONTENT, nextDocumentModel);
    }

    public void setPreviousContent(DocumentModel previousDocumentModel) {
        put(ModelAttributes.PREVIOUS_CONTENT, previousDocumentModel);
    }
}

File: jbake-core/src/main/java/org/jbake/model/DocumentModel.kt
package org.jbake.model

import org.jbake.app.DbUtils
import java.util.*

class DocumentModel : BaseModel() {

    var body: String
        get() = get(ModelAttributes.DOC_BODY_RENDERED) as String
        set(body) { put(ModelAttributes.DOC_BODY_RENDERED, body) }

    var date: Date?
        get() = get(ModelAttributes.DOC_DATE) as Date?
        set(date) = putOrRemoveIfNull(ModelAttributes.DOC_DATE, date)

    var status: String?
        get() = if (containsKey(ModelAttributes.DOC_STATUS)) get(ModelAttributes.DOC_STATUS) as String? else ""
        set(status) = putOrRemoveIfNull(ModelAttributes.DOC_STATUS, status)

    var type: String
        get() = if (containsKey(ModelAttributes.DOC_TYPE)) get(ModelAttributes.DOC_TYPE) as String else ""
        set(type) { put(ModelAttributes.DOC_TYPE, type) }

    var tags: List<String>
        get() = DbUtils.toStringList(get(ModelAttributes.DOC_TAGS) ?: emptyList<String>())
        set(tags) { put(ModelAttributes.DOC_TAGS, tags) }

    var sha1: String?
        get() = get(ModelAttributes.FS_DOC_SHA1) as String?
        set(sha1) { putOrRemoveIfNull(ModelAttributes.FS_DOC_SHA1, sha1) }

    var sourceUri: String?
        get() = get(ModelAttributes.FS_DOC_SOURCE_REL_URI) as String?
        set(uri) { putOrRemoveIfNull(ModelAttributes.FS_DOC_SOURCE_REL_URI, uri) }

    var rootPath: String
        get() = get(ModelAttributes.FS_REL_FROM_DOC_TO_SITEROOT) as String
        set(pathToRoot) { put(ModelAttributes.FS_REL_FROM_DOC_TO_SITEROOT, pathToRoot) }

    var rendered: Boolean
        get() = getOrDefault(ModelAttributes.FS_DOC_WAS_RENDERED, false) as Boolean
        set(rendered) { put(ModelAttributes.FS_DOC_WAS_RENDERED, rendered) }

    var file: String?
        get() = get(ModelAttributes.FS_DOC_SOURCE_PATH_ABS) as String?
        set(path) { putOrRemoveIfNull(ModelAttributes.FS_DOC_SOURCE_PATH_ABS, path) }

    var noExtensionUri: String?
        get() = get(ModelAttributes.FS_DOC_OUTPUT_URI_NOEXT) as String?
        set(noExtensionUri) { putOrRemoveIfNull(ModelAttributes.FS_DOC_OUTPUT_URI_NOEXT, noExtensionUri) }

    var title: String?
        get() = get(ModelAttributes.DOC_TITLE) as String?
        set(title) { putOrRemoveIfNull(ModelAttributes.DOC_TITLE, title) }

    var cached: Boolean?
        get() {
            val value = get(ModelAttributes.FS_DOC_IS_CACHED_IN_DB)
            return if (value is String) value.toBoolean() else value as Boolean?
        }
        set(cached) { putOrRemoveIfNull(ModelAttributes.FS_DOC_IS_CACHED_IN_DB, cached) }

    var nextContent: DocumentModel?
        get() = get(ModelAttributes.PAGI_NEXT_CONTENT) as DocumentModel?
        set(nextDocumentModel) { putOrRemoveIfNull(ModelAttributes.PAGI_NEXT_CONTENT, nextDocumentModel) }

    var previousContent: DocumentModel?
        get() = get(ModelAttributes.PAGI_PREV_CONTENT) as DocumentModel?
        set(previousDocumentModel) { putOrRemoveIfNull(ModelAttributes.PAGI_PREV_CONTENT, previousDocumentModel) }


    companion object {
        @JvmStatic
        fun createDefaultDocumentModel(): DocumentModel = DocumentModel().apply {
            cached = true
            rendered = false
        }
    }
}
